### ISSUE TITLE
Timeouts under high load with 2.1.17 and 2.1.18 

### ISSUE THREAD

**Issue by jsyrjala** [http-kit/http-kit#148]
http-kit versions 2.1.17 and 2.1.18 have rare, random timeouts with high load. Version 2.1.16 and earlier work without timeouts. 

Tests are run against this code https://github.com/jsyrjala/http-kit-perf . Tested with [ab](http://httpd.apache.org/docs/2.2/programs/ab.html)

Failing ab runs looks like this. Timeouts don't happen on every run, but run it  a few times, and you should see this behavior.

```
ab -n 10000 -c 100 http://localhost:9876/                      
This is ApacheBench, Version 2.3 <$Revision: 1554214 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking localhost (be patient)
Completed 1000 requests
Completed 2000 requests
Completed 3000 requests
Completed 4000 requests
Completed 5000 requests
Completed 6000 requests
Completed 7000 requests
Completed 8000 requests
Completed 9000 requests
apr_pollset_poll: The timeout specified has expired (70007)
Total of 9994 requests completed
```

**Comment by nberger** [comment]
With the help of `git bisect` I tracked this issue down to 3e9d112. That's the commit that introduced this issue IMO.

**Comment by ddossot** [comment]
@nberger :bow:

**Comment by ptaoussanis** [comment]
@nberger: Thanks Nicolás!

**Comment by liuchong** [comment]
seems the version http-kit "2.1.21-alpha2" still has this issue.

**Comment by ptaoussanis** [comment]
@liuchong Yes, thank you for the confirmation. @nberger helped tracked down the problem to commit https://github.com/http-kit/http-kit/commit/3e9d112f3db8264d61f5713b6bfbb626b05066d3. **PR to fix welcome**, otherwise will need to wait on a response from http-kit's author @shenfeng.

**Comment by nberger** [comment]
In trying to understand what's the reasoning behind 3e9d112, I checked the issues activity around the date of the commit and it seems these are the ones related to it:

- #105: HTTPServer.stop is not threadsafe.
- #73: Exceptions when server is under load

Then I tried to split this commit into smaller commits that seem to map to those issues:

- #105 maps to https://github.com/nberger/http-kit/commit/4e65b1c966e8ed25efe95c15a79342f92f348bd3
- #73 maps to  https://github.com/nberger/http-kit/commit/25c2fe8374d15c22c6c5adf14c885f02f71e095a

There's also https://github.com/nberger/http-kit/commit/ce241297f2c6756776d0fce4f5e753a67bc24087 but doesn't seem to be so important. Would be nice to understand why it is there though :)

Reverting the commit related to #73 makes this timeout to go away. It can be reverted cleanly on master, I did exactly that in https://github.com/nberger/http-kit/tree/fix-148. But I guess it'd make #73 to come back in all its fury :).

So I think we need someone to dig deep into #73 and its related commit to debug what kind of race condition is happening and how to avoid it.
